teuthology/task/ansible.py

```diff
@@ -0,0 +1,225 @@
+        self.logger = logger
+        self.level = level
+
+    def write(self, string):
+        self.logger.log(self.level, string)
+
+    def flush(self):
+        pass
+
+
+class Ansible(Task):
+    """
+    A task to run ansible playbooks
+
+    Required configuration parameters:
+        playbook:   Required; can either be a list of plays, or a path/URL to a
+                    playbook
+
+    Optional configuration parameters:
+        repo:       A path or URL to a repo (defaults to '.'). Given a repo
+                    value of 'foo', ANSIBLE_ROLES_PATH is set to 'foo/roles'
+        hosts:      A list of teuthology roles or partial hostnames (or a
+                    combination of the two). ansible-playbook will only be run
+                    against hosts that match.
+        inventory:  A path to be passed to ansible-playbook with the
+                    --inventory-file flag; useful for playbooks that also have
+                    vars they need access to. If this is not set, we check for
+                    /etc/ansible/hosts and use that if it exists. If it does
+                    not, we generate a temporary file to use.
+
+    Examples:
+
+    tasks:
+    - ansible:
+        repo: https://github.com/ceph/ceph-cm-ansible.git
+        playbook:
+          - roles:
+            - some_role
+            - another_role
+        hosts:
+          - client.0
+          - host1
+
+    tasks:
+    - ansible:
+        repo: /path/to/repo
+        inventory: /path/to/inventory
+        playbook: /path/to/playbook.yml
+        tags: my_tags
+
+    """
+    def __init__(self, ctx, config):
+        super(Ansible, self).__init__(ctx, config)
+        self.log = log
+        self.generated_inventory = False
+        self.generated_playbook = False
+
+    def setup(self):
+        super(Ansible, self).setup()
+        self.find_repo()
+        self.get_playbook()
+        self.get_inventory() or self.generate_hosts_file()
+        if not hasattr(self, 'playbook_file'):
+            self.generate_playbook()
+
+    def find_repo(self):
+        """
+        Locate the repo we're using; cloning it from a remote repo if necessary
+        """
+        repo = self.config.get('repo', '.')
+        if repo.startswith('http://') or repo.startswith('https://') or \
+                repo.startswith('git@'):
+            repo_path = fetch_repo(repo,
+                                   self.config.get('branch', 'master'),
+                                   )
+        else:
+            repo_path = os.path.abspath(os.path.expanduser(repo))
+        self.repo_path = repo_path
+
+    def get_playbook(self):
+        """
+        If necessary, fetch and read the playbook file
+        """
+        playbook = self.config['playbook']
+        if isinstance(playbook, list):
+            # Multiple plays in a list
+            self.playbook = playbook
+        elif isinstance(playbook, str) and (playbook.startswith('http://') or \
```

Comment:
nitpick: checking again here to see if something is a url.  Maybe a utility method for these types of checks would be worth it?